contracts/core/src/solve/Inbox.sol

```diff
@@ -0,0 +1,114 @@
+
+    /**
+     * @dev Last request ID assigned.
+     */
+    uint256 internal _lastId;
+
+    /**
+     * @notice Map ID to request.
+     */
+    mapping(uint256 id => Solve.Request) internal _requests;
+
+    function initialize() public initializer {
+        __ReentrancyGuard_init();
+    }
+
+    /**
+     * @notice Returns the request with the given ID.
+     */
+    function getRequest(uint256 id) external view returns (Solve.Request memory) {
+        return _requests[id];
+    }
+
+    /**
+     * @notice Open a request to execute a call on another chain, backed by deposits.
+     *  ERC20 deposits are transferred from msg.sender from to this inbox. Native deposit must match msg.value.
+     * @param call      Details of the call to be executed on another chain.
+     * @param deposits  Array of deposits backing the request.
+     */
+    function request(Solve.Call calldata call, Solve.TokenDeposit[] calldata deposits)
+        external
+        payable
+        nonReentrant
+        returns (uint256 id)
+    {
+        if (call.target == address(0)) revert InvalidCall();
+        if (call.destChainId == 0) revert InvalidCall();
+        if (call.data.length == 0) revert InvalidCall();
+        if (deposits.length == 0 && msg.value == 0) revert NoDeposits();
+
+        Solve.Request storage req = _openRequest(msg.sender, call, deposits);
+
+        emit Requested(req.id, req.from, req.call, req.deposits);
+
+        return req.id;
+    }
+
+    /**
+     * @dev Open a new request in storage at `id`.
+     *      Transfer token deposits from msg.sender to this inbox.
+     *      Duplicate token addresses are allowed.
+     */
+    function _openRequest(address from, Solve.Call calldata call, Solve.TokenDeposit[] calldata deposits)
+        internal
+        returns (Solve.Request storage req)
+    {
+        uint256 id = _nextId();
+
+        req = _requests[id];
+        req.id = id;
+        req.openedAt = uint96(block.timestamp);
+        req.from = from;
+        req.call = call;
+
+        if (msg.value > 0) {
+            req.deposits.push(Solve.Deposit({ isNative: true, token: address(0), amount: msg.value }));
+        }
+
+        for (uint256 i = 0; i < deposits.length; i++) {
+            if (deposits[i].amount == 0) revert InvalidDeposit();
+            if (deposits[i].token == address(0)) revert InvalidDeposit();
+
+            req.deposits.push(Solve.Deposit({ isNative: false, token: deposits[i].token, amount: deposits[i].amount }));
+
+            // NOTE: all external methods must be nonReentrant
+            // This allows us to transfer while opening the request - saving some gas.
+            IERC20(deposits[i].token).safeTransferFrom(msg.sender, address(this), deposits[i].amount);
+        }
+    }
+
+    /**
+     * @dev Increment and return _lastId.
+     */
+    function _nextId() internal returns (uint256) {
+        _lastId++;
+        return _lastId;
+    }
```

Comment:
@Zodomo what was argument for bytes32 w/ assembly increment you used in prototype again?

Reply:
```solidity
assembly {
    // Load from storage slot
    guid := sload(nextGuid.slot)
    // Increment and store back
    sstore(nextGuid.slot, add(guid, 1))
}
```

Where `nextGuid` is a storage variable

Reply:
yes - what was the motivation? why better than just `uint256` and plain ol sol

Reply:
Ah, bytes32 values are smaller and easier to pass around. They visually "look" more like a guid. Just a style choice tbh. We can also opt for truly random guid's that are always 64 characters rather than unwieldly large numbers that can far outpace a bytes32 in text length.

Reply:
`115792089237316195423570985008687907853269984665640564039457584007913129639935`
`0xffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff`

Purely a style decision imo. The assembly is just to increment it without wasting gas on casting.

Reply:
bytes32 and uint256 are both 32 bytes not?

Reply:
They are more visually pleasing, indeed

Reply:
I'll switch to bytes32 and then we can have a look at gas savings if we do assmbly bit